include/secp256k1_frost.h

```diff
@@ -1,17 +1,110 @@
+ *  Args:    ctx: a secp256k1 context object
+ *  Out:   out32: pointer to a 32-byte array to store the serialized share
+ *  In:    share: pointer to the share
+ */
+SECP256K1_API int secp256k1_frost_share_serialize(
+    const secp256k1_context* ctx,
+    unsigned char *out32,
+    const secp256k1_frost_share* share
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Parse a FROST share.
+ *
+ *  Returns: 1 when the share could be parsed, 0 otherwise.
+ *  Args:    ctx: a secp256k1 context object
+ *  Out:   share: pointer to a share object
+ *  In:     in32: pointer to the 32-byte share to be parsed
+ */
+SECP256K1_API int secp256k1_frost_share_parse(
+    const secp256k1_context* ctx,
+    secp256k1_frost_share* share,
+    const unsigned char *in32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Creates a key generation share and verifiable secret sharing ("VSS")
+ *  commitments.
+ *
+ *  To generate a key, each participant generates a share for each other
+ *  participant. For example, in the case of 2 particpants, Alice and Bob, they
+ *  each generate 2 shares, distribute 1 share to each other using a secure
+ *  channel, and keep 1 for themselves.
+ *
+ *  Each participant _must_ have a secure channel with each other participant
+ *  with which they can transmit shares to each other.
+ *
+ *  A new session_id32 _must_ be used for each key generation session. For
+ *  example, in the case of 2 participants, Alice and Bob, Alice will generate
+ *  a session_id32 and use it for each of the 2 calls to
+ *  secp256k1_frost_share_gen and Bob will generate a session_id32 and use it
+ *  for each of the 2 calls to secp256k1_frost_share_gen. Both Alice and Bob
+ *  must NOT REUSE there respective session_id32 again for subsequent key
+ *  generation sessions. If Alice and Bob fail to complete this session or
+ *  start a new session to generate a new key, they must NOT REUSE their
+ *  respective session_id32 again, but instead generate a new one. It is
+ *  recommended to always choose session_id32 uniformly at random to avoid
+ *  their reuse.
+ *
+ *  Returns: 0 if the arguments are invalid, 1 otherwise
+ *  Args:            ctx: pointer to a context object initialized for
+ *                        verification
+ *  Out:  vss_commitment: the coefficient commitments. The length of this array
+ *                        must be equal to the threshold (can be NULL).
+ *                 share: pointer to the key generation share
+ *   In:    session_id32: a 32-byte session_id32 as explained above
+ *               keypair: pointer to a keypair that contains the secret that is
+ *                        shared
+ *                    pk: pointer to the public key of the share recipient
+ *             threshold: the minimum number of signers required to produce a
+ *                        signature
+ */
+SECP256K1_API int secp256k1_frost_share_gen(
```

Comment:
Also, it seems like the PoK of the first coefficient(secret key) is missing? (step 2 of the figure) this is required to prevent a t-1 rouge key attack (we could also maybe use a MuSig-like trick but we might need a security proof for feldman + musig trick @real-or-random )

Reply:
>Also, if I understand this correctly, each participant P_i should run this function should in a loop over all P_j (i != j) participants? isn't that a bit wasteful?(regenerating all the coefficients) why not pass in a list of pks, and output a state + all messages

For devices with limited resources, the API provides the flexibility to generate the data sequentially instead of simultaneously, and it only needs to store its private key and session_id (because shares are generated deterministically from those values) and can go offline between share generations if it doesn't have all the public keys for each participant when it begins generating shares.

The commitments are an optional argument to avoid requiring that they be regenerated over and over again, but it's a good point that the coefficients get regenerated each time.

@l2xl also makes the [suggestion](https://github.com/ElementsProject/secp256k1-zkp/pull/138#issuecomment-1201855897) for even more API flexibility:

>Please split the secp256k1_frost_share_gen() function in two:
> * secp256k1_frost_keyshare_commitment_gen()
> * secp256k1_frost_keyshare_gen()

Reply:
I'm still unsure if sending the evaluation of `f(index)` together with the coefficients is actually safe. this reduces a round from the FROST keygen algorithm so seems like something they'd want to do in the paper if safe.

I'd need to go over their security proof and see the reasoning there to see if this is used, if you have any insights / justifications I'd love to hear

Reply:
@elichai That's a good point. This optimization doesn't even save us a round, it just saves sending an extra signature.

In the prescribed protocol in this PR, there's 2 rounds:

1. distribute shares and coefficient commitments
2. distribute signatures of VSS hashes

If we stick to the paper (and add VSS hash sigs for the broadcast requirement), there's also 2 rounds:

1. distribute coefficient commitments and signatures of a context string
2. distribute shares and signatures of the VSS hash

With the paper there's 2 distributions of signatures, the first signs a context string to prevent a rogue key attack and the second signs the VSS hash for the broadcast requirement.

It would only take some slight revisions to the docs and maybe the APIs as well to support this change. The optimization is probably not worth the time and effort required to prove it secure so I'm currently thinking I should make this change.

Reply:
Hmm you're right. the "3rd round" in my head was actually not a round but a local computation.

I'll try to figure out from the proof in the paper if it's safe to send the evaluation before first committing to the coefficients + the zkprof of ai_0

did you write up something about this signature satisfying the broadcast requirement?

Reply:
>did you write up something about this signature satisfying the broadcast requirement?

The only reference to it is [here](https://github.com/ElementsProject/secp256k1-zkp/blame/b0234a23eed456dc0b63a3a9cbc2e6ba68588bb3/src/modules/frost/frost.md#L23-L24).

In Section 2.1 of the FROST paper, it states:

>In practice, implementations guarantee consistency of participants’ views by using techniques such as posting commitments to a centralized server that is trusted to provide a single view to all participants, or adding another protocol round where participants compare their received commitment values to ensure they are identical.

The VSS hash signatures are intended to serve as that comparison of received commitment values.